# HR Promotion clasification Problem

Your client is a large MNC and they have 9 broad verticals across the organisation. One of the problem your client is facing is around identifying the right people for promotion (only for manager position and below) and prepare them in time. Currently the process, they are following is:

- They first identify a set of employees based on recommendations/ past performance
- Selected employees go through the separate training and evaluation program for each vertical. These programs are based on the required skill of each vertical
- At the end of the program, based on various factors such as training performance, KPI completion (only employees with KPIs completed greater than 60% are considered) etc., employee gets promotion

For above mentioned process, the final promotions are only announced after the evaluation and this leads to delay in transition to their new roles. Hence, company needs your help in identifying the eligible candidates at a particular checkpoint so that they can expedite the entire promotion cycle.

- Attribute Information:

- employee_id	Unique ID for employee
- department	Department of employee
- region	Region of employment (unordered)
- education	Education Level
- gender	Gender of Employee
- recruitment_channel	Channel of recruitment for employee
- no_of_trainings	no of other trainings completed in previous year on soft - skills, technical skills etc.
- age	Age of Employee
- previous_year_rating	Employee Rating for the previous year
- length_of_service	Length of service in years
- KPIs_met >80%	if Percent of KPIs(Key performance Indicators) >80% then 1 else 0
- awards_won?	if awards won during previous year then 1 else 0
- avg_training_score	Average score in current training evaluations
- is_promoted	(Target) Recommended for promotion

# Importing required libraries for the project

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

- Loading Performence dataset from git hub account

In [2]:
Perfdf = pd.read_csv('https://raw.githubusercontent.com/Manju410/MLPractice/main/data/Promotion_Cleanup.csv')

In [3]:
Perfdf.head()

education  gender  ...  recchan_referred  recchan_sourcing
0          2       0  ...                 0                 1
1          1       1  ...                 0                 0
2          1       1  ...                 0                 1
3          1       1  ...                 0                 0
4          1       1  ...                 0                 0

[5 rows x 56 columns]

- Number of rows and columns in the dataset

In [4]:
Perfdf.shape

(54808, 56)

- There are 54808 rows and 56 columns in the above dataset

- Information about dataset like datatype,count etc

In [5]:
Perfdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54808 entries, 0 to 54807
Data columns (total 56 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   education                 54808 non-null  int64  
 1   gender                    54808 non-null  int64  
 2   no_of_trainings           54808 non-null  int64  
 3   age                       54808 non-null  int64  
 4   previous_year_rating      54808 non-null  float64
 5   length_of_service         54808 non-null  int64  
 6   KPI_Score                 54808 non-null  int64  
 7   awards_won                54808 non-null  int64  
 8   avg_training_score        54808 non-null  int64  
 9   is_promoted               54808 non-null  int64  
 10  depart_Analytics          54808 non-null  int64  
 11  depart_Finance            54808 non-null  int64  
 12  depart_HR                 54808 non-null  int64  
 13  depart_Legal              54808 non-null  int64  
 14  depart

- Checking null values in dataset

In [6]:
Perfdf.isna().sum()

education                   0
gender                      0
no_of_trainings             0
age                         0
previous_year_rating        0
length_of_service           0
KPI_Score                   0
awards_won                  0
avg_training_score          0
is_promoted                 0
depart_Analytics            0
depart_Finance              0
depart_HR                   0
depart_Legal                0
depart_Operations           0
depart_Procurement          0
depart_R&D                  0
depart_Sales & Marketing    0
depart_Technology           0
reg_region_1                0
reg_region_10               0
reg_region_11               0
reg_region_12               0
reg_region_13               0
reg_region_14               0
reg_region_15               0
reg_region_16               0
reg_region_17               0
reg_region_18               0
reg_region_19               0
reg_region_2                0
reg_region_20               0
reg_region_21               0
reg_region

# K Neigbours clasifier
 
- 0- is not promoted
- 1- is promoted

In [7]:
X = Perfdf.drop('is_promoted',axis=1)
y = Perfdf.is_promoted
X.shape, y.shape

((54808, 55), (54808,))

In [8]:
from sklearn.model_selection import train_test_split

In [9]:
X_trian,X_test,y_train,y_test = train_test_split(X,y,test_size=0.25,random_state=1)

In [10]:
X_trian.shape , X_test.shape , y_train.shape, y_test.shape

((41106, 55), (13702, 55), (41106,), (13702,))

In [13]:
from sklearn.neighbors import KNeighborsClassifier

In [14]:
Kneighbors = KNeighborsClassifier()

In [15]:
Kneighbors.fit(X_trian,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [16]:
y_learnpred = Kneighbors.predict(X_trian)

In [17]:
from sklearn.metrics import confusion_matrix,classification_report

In [18]:
confusion_matrix(y_train,y_learnpred)

array([[37515,   100],
       [ 2739,   752]])

- here TP,TN,FP,FN are given in matrix
- 37515 is True Negative: predicted correctly as not promoted employees
- 752 is True Positive: predicted correctly as promoted employees
- 2739 is False Negative: predicted wrongly as not promoted employees
- 100 is False Positive: predicted wrongly as promoted employees

# so here False Positive(100) is importent because model predicting not promoted employees as predicted

In [20]:
y_testpred = Kneighbors.predict(X_test)

In [21]:
confusion_matrix(y_test,y_testpred)

array([[12438,    87],
       [ 1002,   175]])

In [22]:
print(classification_report(y_test,y_testpred,target_names=['not-promoted','promoted']))

              precision    recall  f1-score   support

not-promoted       0.93      0.99      0.96     12525
    promoted       0.67      0.15      0.24      1177

    accuracy                           0.92     13702
   macro avg       0.80      0.57      0.60     13702
weighted avg       0.90      0.92      0.90     13702

